microceph/ceph/maintenance.go

```diff
@@ -0,0 +1,297 @@
+package ceph
+
+import (
+	"fmt"
+
+	"github.com/canonical/lxd/shared/logger"
+
+	microCli "github.com/canonical/microcluster/v2/client"
+
+	"github.com/canonical/microceph/microceph/client"
+)
+
+// EnterMaintenance put a given node into maintanence mode.
+func EnterMaintenance(clusterClient *microCli.Client, cephClient client.ClientInterface, name string, force, dryRun, setNoout, stopOsds bool) error {
+	ops := []operation{}
+
+	// pre-flight checks
+	if !force {
+		ops = append(ops, []operation{
+			&checkNodeInClusterOps{cephClient, clusterClient},
+			&checkOsdOkToStopOps{cephClient, clusterClient},
+			&checkNonOsdSvcEnoughOps{cephClient, clusterClient, 3, 1, 1},
+		}...)
+	}
+
+	// optionally set noout
+	if setNoout {
+		ops = append(ops, []operation{
+			&setNooutOps{},
+			&assertNooutFlagSetOps{},
+		}...)
+	}
+
+	// optionally stop osd service
+	if stopOsds {
+		ops = append(ops, []operation{
+			&stopOsdOps{cephClient, clusterClient},
+		}...)
+	}
+
+	m := maintenance{name}
+	err := m.Run(ops, dryRun)
+	if err != nil {
+		return fmt.Errorf("Failed to enter maintenance mode: %v", err)
+	}
+	return nil
+}
+
+// ExitMaintenance recover a given node from maintanence mode.
+func ExitMaintenance(clusterClient *microCli.Client, cephClient client.ClientInterface, name string, dryRun bool) error {
+	ops := []operation{}
+
+	// preflight checks
+	ops = append(ops, []operation{
+		&checkNodeInClusterOps{cephClient, clusterClient},
+	}...)
+
+	// idempotently unset noout and start osd service
+	ops = append(ops, []operation{
+		&unsetNooutOps{},
+		&assertNooutFlagUnsetOps{},
+		&startOsdOps{cephClient, clusterClient},
+	}...)
+
+	m := maintenance{name}
+	err := m.Run(ops, dryRun)
+	if err != nil {
+		return fmt.Errorf("Failed to exit maintenance mode: %v", err)
+	}
+	return nil
+}
+
+type maintenance struct {
+	nodeName string
+}
+
+func (m *maintenance) Run(operations []operation, dryRun bool) error {
+	for _, ops := range operations {
+		if dryRun {
+			fmt.Println(ops.DryRun(m.nodeName))
+		} else {
+			err := ops.Run(m.nodeName)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
```

Comment:
**suggestion:** 

*maintenance* struct seems not necessary here.
I think it can be merge to one simple function *RunOperations*

Reply:
Can be resolved.